### PR TITLE
Refactor transaction into batch and snapshot

### PR DESCRIPTION
**Pull request by weiihann**
### Description
This PR removes the Transaction struct and added `batch` and `snapshot` which implements the `Transaction` interface. `batch` is used for read-write operations while `snapshot` is used for read-only operations.

### Rationale
A transaction can only be a `batch` or a `snapshot`, not both at the same time. The following pattern may be unnecessary:
```
if t.batch != nil {
  ...do something with batch
} else if t.snapshot != nil {
  ...do something with snapshot 
}
```

It's clearer to just separate `batch` and `snapshot` into different structs and use them accordingly.

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -77,9 +77,6 @@ type Transaction interface {
 	// Get fetches the value for the given key, should return ErrKeyNotFound if key is not present
 	// Caller should not assume that the slice would stay valid after the call to cb
 	Get(key []byte, cb func([]byte) error) error
-
-	// Impl returns the underlying transaction object
-	Impl() any
```
Comment (kirugan) [inline db/db.go:82]: @IronGauntlets do you remember why this method was introduced in the first place? how is this different from `db.(customType)` ?

File: db/pebble/batch.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// Commit : see db.Transaction.Commit
+func (b *batch) Commit() error {
+	start := time.Now()
+	defer func() { b.listener.OnCommit(time.Since(start)) }()
+
+	if b.batch == nil {
+		return ErrDiscardedTransaction
+	}
```
Comment (kirugan) [inline db/pebble/batch.go:50]: let's move this check up to beginning of this method (same as in `Discard`)

Comment (kirugan) [reply db/pebble/batch.go:50]: time difference that is passed to `b.listener.OnCommit` is negligible

**Review by kirugan — APPROVED** [review]

File: db/db.go (hunk 1 above)
Comment (IronGauntlets) [reply db/db.go:82]: I wasn't too sure about this when Omer implemented it. If it is not used anywhere we can delete it.

File: db/pebble/db.go [hunk 3]
```diff
@@ -18,6 +19,10 @@ const (
 	minCacheSizeMB = 8
 )
 
+var (
+	ErrDiscardedTransaction = errors.New("discarded txn")
+	ErrReadOnlyTransaction  = errors.New("read-only transaction")
```
Comment (IronGauntlets) [inline db/pebble/db.go:24]: Why the disparity betweeen `txn` vs `transaction`?

File: db/pebble/common.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	defer listener.OnIO(false, time.Since(start)) //nolint:govet
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return db.ErrKeyNotFound
+		}
+		return err
+	}
+
+	return utils.RunAndWrapOnError(closer.Close, cb(val))
+}
```
Comment (IronGauntlets) [inline db/pebble/common.go:34]: Would a generic be a better idea?

**Review by IronGauntlets — APPROVED** [review]
In general, it looks good to me. I have let some comments

File: db/pebble/common.go (hunk 4 above)
Comment (kirugan) [reply db/pebble/common.go:34]: Tbh I don't think generic will bring a lot of value. The code will become a little bit harder to read

**Pull request merged**